feat: Admin Dashboard

krantheman opened the pull request:
## Domains

<details>
<summary>Screenshots</summary>

### List

![image](https://github.com/user-attachments/assets/5a0fad8b-d0fd-4e92-b183-9bf377abb997)

### Add Domain

![image](https://github.com/user-attachments/assets/f12e804f-4d0c-4418-908f-9dcbf07b2f8e)

![image](https://github.com/user-attachments/assets/6fe97439-347a-497e-8746-fcd42b902181)

### Domain Settings

![image](https://github.com/user-attachments/assets/85a5feaa-f96c-442f-82a1-0361c6052da1)

</details>


## Members

<details>
<summary>Screenshots</summary>

### List

![image](https://github.com/user-attachments/assets/1bf540f7-a816-4d8a-987d-1a16c6c94bc3)

### Add Member

![image](https://github.com/user-attachments/assets/707d86d5-e824-4ed5-8b1d-917f2815837c)

![image](https://github.com/user-attachments/assets/9e75c2f1-2027-4687-807e-685140e12033)

</details>

## Mail Account Request

- Role (Select) field -> Is Admin (Checkbox)
- Form UX
- Validations

## Other

- Upgrade frappe-ui to v0.1.108
- Downgrade vite to v5 (frappe-ui compatibility)
- Fix email css build
- Routing
	- Administrator -> /desk
	- Mail Admin w/o account -> /mail/dashboard
	- Mail Admin w/ account -> /mail/inbox
	- Mail User w/ account-> /mail/inbox 
	- Mail User w/o account -> ??

krantheman committed:
feat: add Domains page

krantheman committed:
feat: redirect non admins out of dashboard

krantheman committed:
chore: clean up redundant 'computed's

krantheman committed:
chore: move admin methods to different file

krantheman committed:
feat: add Add Domain dialog

krantheman committed:
feat: add domain listview

krantheman committed:
chore: merge with develop

krantheman committed:
fix: domain request apis

krantheman committed:
feat: add single domain view

krantheman committed:
chore: bump frappe-ui to v0.1.108

krantheman committed:
feat(Mail Domain): add tenant name

krantheman committed:
feat: show mail domain doc fields

s-aga-r commented on mail/mail/doctype/mail_domain/mail_domain.json:
```diff
@@ -166,6 +167,14 @@
    "reqd": 1,
    "search_index": 1
+  },
+  {
+   "fetch_from": "tenant.tenant_name",
+   "fieldname": "tenant_name",
+   "fieldtype": "Data",
+   "hidden": 1,
+   "label": "Tenant Name",
+   "read_only": 1
   }
```
Why do we need `tenant_name` in the Mail Domain?

krantheman replied:
For fetching tenant name for the ui, instead of having to query it repeatedly

([Commonly done in hrms for employee_name](https://github.com/frappe/hrms/blob/c141c7df600af68b9d9619e6852c1692d71f6bfb/hrms/hr/doctype/appraisal/appraisal.json#L71))

s-aga-r replied:
> For fetching tenant name for the ui, instead of having to query it repeatedly
> 
> ([Commonly done in hrms for employee_name](https://github.com/frappe/hrms/blob/c141c7df600af68b9d9619e6852c1692d71f6bfb/hrms/hr/doctype/appraisal/appraisal.json#L71))

`tenant_name` is editable in Mail Tenant, what if the user changes it?

![image](https://github.com/user-attachments/assets/5f090f86-ffbb-4985-86f2-dcf438bb17bc)

krantheman replied:
You're right; [reverted](https://github.com/frappe/mail/pull/68/commits/f97c47035ac93d7fee9e2f79ffce28761ff0f170).

s-aga-r commented on mail/api/admin.py:
```diff
@@ -0,0 +1,32 @@
+
+
+@frappe.whitelist()
+def create_domain_request(domain_name, mail_tenant) -> str:
```
Change the return type to "Mail Domain Request" or the return value.

krantheman committed:
fix(Signup): account request value

krantheman committed:
chore: add frappe-ui.diff

krantheman committed:
chore: downgrade vite to v5

krantheman committed:
feat(Domains): add styling to listview

krantheman committed:
feat(Domain): show error on setvalue failure

krantheman committed:
feat(Domain): add labels

krantheman committed:
refactor(Domain): save doc using button

krantheman committed:
feat(Domain): add actions resources

krantheman committed:
feat(Domain): add confirmation dialog

s-aga-r commented on mail/mail/doctype/mail_domain/mail_domain.py:
```diff
@@ -126,7 +126,7 @@ def refresh_dns_records(self, do_not_save: bool = False) -> None:
 
 	@frappe.whitelist()
-	def verify_dns_records(self, do_not_save: bool = False) -> None:
+	def verify_dns_records(self, do_not_save: bool = False) -> 1 | 0:
```
```suggestion
	def verify_dns_records(self, do_not_save: bool = False) -> bool:
```

s-aga-r commented on mail/mail/doctype/mail_domain/mail_domain.py:
```diff
@@ -151,6 +151,8 @@ def verify_dns_records(self, do_not_save: bool = False) -> None:
 		if not do_not_save:
 			self.save(ignore_permissions=True)
 
+		return self.is_verified
```
```suggestion
		return bool(self.is_verified)
```

krantheman committed:
fix(Domain): rewrite of dns records on save

krantheman committed:
fix(Domain): column widths

krantheman committed:
fix(Domain): isDirty Save disable

krantheman committed:
chore: fix return types

s-aga-r commented on mail/api/admin.py:
```diff
@@ -0,0 +1,34 @@
+import frappe
+
+from mail.mail.doctype.mail_domain_request.mail_domain_request import MailDomainRequest
```
```suggestion
from typing import TYPE_CHECKING
if TYPE_CHECKING:
	from mail.mail.doctype.mail_domain_request.mail_domain_request import MailDomainRequest
```

s-aga-r commented on mail/api/admin.py:
```diff
@@ -0,0 +1,34 @@
+
+
+@frappe.whitelist()
+def create_domain_request(domain_name, mail_tenant) -> MailDomainRequest:
```
```suggestion
def create_domain_request(domain_name, mail_tenant) -> "MailDomainRequest":
```

krantheman committed:
chore: fix type checking

krantheman committed:
chore: merge with develop

s-aga-r commented on mail/mail/doctype/mail_domain/mail_domain.json:
```diff
@@ -165,8 +166,15 @@
    "reqd": 1,
-   "search_index": 1,
-   "set_only_once": 1
+   "search_index": 1
+  },
```
```suggestion
   "search_index": 1,
   "set_only_once": 1
  },
```

krantheman replied:
my bad

krantheman committed:
feat: add Members page

krantheman committed:
chore(Mail Domain): set_only_once as 1

Co-authored-by: s-aga-r <sagarsharma.s312@gmail.com>

krantheman committed:
feat: add Invite User

krantheman committed:
feat: add Invite User functionality

s-aga-r commented on mail/api/account.py:
```diff
@@ -1,21 +1,37 @@
+		account_request.account = f"{username}@{domain}"
+		account_request.invited_by = frappe.session.user
+
 	account_request.insert(ignore_permissions=True)
```
Ignore permissions only for the self-signup (Mail Admin) as only admin(s) can send invitations.

Current case: `allow_guest=True` + `ignore_permissions=True` unauthorized or guest users can send invites for any tenant.

Edit: [Validation](https://github.com/frappe/mail/blob/ee9ffd64b4be39f247fd384907e1fd729442dd93/mail/mail/doctype/mail_account_request/mail_account_request.py#L60-L78) is there but still better to insert without ignoring permissions.

krantheman committed:
feat: show tenant members

krantheman committed:
chore: rename Invite User as Add Member

krantheman committed:
feat: show if admin in members list view

krantheman committed:
chore: merge with develop

krantheman committed:
feat(Members): add actions

krantheman committed:
feat(Members): add non invite based account creation

krantheman committed:
fix: reset state on opening modals

krantheman committed:
refactor: multiple
- move add_member to admin.py
- fix signup link
- show account as email while signing up

krantheman committed:
fix: multiple
- listview empty state
- email of user created

krantheman committed:
refactor: add validation for email address when needed

s-aga-r commented on mail/api/account.py:
```diff
@@ -60,53 +54,26 @@ def get_account_request(request_key: str) -> dict:
 
-	account_request, email, tenant, role = frappe.db.get_value(
-		"Mail Account Request", {"request_key": request_key}, ["name", "email", "tenant", "role"]
+	account_request, email, tenant, role, account = frappe.db.get_value(
+		"Mail Account Request", {"request_key": request_key}, ["name", "email", "tenant", "role", "account"]
 	)
```
For invalid request key:
 
![image](https://github.com/user-attachments/assets/b97a53fd-cf7f-48c6-9b75-79ff98bece2f)

s-aga-r commented on mail/api/account.py:
```diff
@@ -60,53 +54,26 @@ def get_account_request(request_key: str) -> dict:
+		"Mail Account Request", {"request_key": request_key}, ["name", "email", "tenant", "role", "account"]
 	)
 
+	user_email = account if account else email
```
```suggestion
	user_email = account or email
```

s-aga-r commented on mail/api/account.py:
```diff
@@ -60,53 +54,26 @@ def get_account_request(request_key: str) -> dict:
+	user.owner = user_email
 	user.new_password = password
-	user.append_roles(role)
+	user.append_roles("Mail User")
```
Not sure if the same function is called for self-signed. self-signed users should not have the Mail User role. Mail User role is to create a Mail Account, which can't be created for a personal email.

s-aga-r commented on mail/api/account.py:
```diff
@@ -60,53 +54,26 @@ def get_account_request(request_key: str) -> dict:
-
-	doc = frappe.get_doc("Mail Domain Request", domain_request)
-	return doc.verify_and_create_domain(save=True)
+		mail_tenant.add_member(user_email)
```
mail_tenant.add_member(user_email, is_admin=role == "Mail Admin")

s-aga-r commented on mail/mail/doctype/mail_account_request/mail_account_request.py:
```diff
@@ -47,15 +57,26 @@ def validate_role(self) -> None:
+	def validate_non_invite(self) -> None:
+		"""Validates self sign up."""
+
+		if frappe.db.exists("User", self.email):
```
User.name and User.email can be different after https://github.com/frappe/frappe/issues/26189

```suggestion
		if frappe.db.exists("User", {"email": self.email}):
```

s-aga-r commented on mail/mail/doctype/mail_account_request/mail_account_request.py:
```diff
@@ -47,15 +57,26 @@ def validate_role(self) -> None:
+		if self.send_email and not self.email:
+			frappe.throw(_("Email is required to send invite"))
+
+		self.email = self.email.strip().lower()
```
email is optional, self.email can be None.

krantheman replied:
Correct, this validates only for if `send_email` is True. `validate_email_address` does not throw an error if the string is empty.

s-aga-r replied:
Works although we should use this [validate_email_address](https://github.com/frappe/mail/blob/d82cb22c5b252d1b6a48e0a74416de06aaf57717/mail/utils/validation.py#L84-L89) for an RCPT check. Should be based on a setting as some self-host users will not like this. Will add it later.

Edit: 

`self.email = self.email.strip().lower()` can raise `AttributeError: 'NoneType' object has no attribute 'strip'`

krantheman replied:
Added [fix](https://github.com/frappe/mail/pull/68/commits/1c23ae609608e88650f3734791a06bd761cafa45)

krantheman committed:
refactor: create mail account

krantheman committed:
fix: email css build

krantheman committed:
feat(Mail Account Request): validate expired

krantheman committed:
fix: validate email only when there is one

krantheman committed:
refactor(Mail Domain): remove tenant_name field

krantheman committed:
Merge branch 'develop' into feat-admin-dashboard

krantheman committed:
fix: frappe-logo typo 💀

krantheman committed:
refactor(Members): only show badge for admins

s-aga-r commented:
- Only show the subdomain and root domain when true.
   ![image](https://github.com/user-attachments/assets/ce0af322-1ba7-43ba-82a9-734600a6604c)

s-aga-r commented on mail/api/admin.py:
```diff
@@ -0,0 +1,84 @@
+@frappe.whitelist()
+def get_tenant_members(tenant: str) -> list[dict[str, str]]:
+	"""Returns list of members for the given tenant"""
+
```
if not is_tenant_admin(tenant, frappe.session.user):
    throw

Currently, any logged-in user can get the list of members for any tenant.

Additionally, the return value can be `[]` if no member exists.

s-aga-r commented on mail/mail/doctype/mail_account_request/mail_account_request.py:
```diff
@@ -15,21 +15,29 @@
 	is_valid_email_for_domain,
 	validate_domain_is_enabled_and_verified,
 	validate_domain_owned_by_tenant,
+	validate_email_address,
```
This wasn't what I meant. Use the `frappe.utils.validate_email_address` only for now. `mail.utils.validation.validate_email_address` can be used to block an invalid email address as it connects to an SMTP server and does an RCPT check to make sure the email is deliverable, which will break signup on a local setup. I will make it configurable somewhere when to use which.

s-aga-r commented on mail/mail/doctype/mail_account_request/mail_account_request.py:
```diff
@@ -15,21 +15,29 @@
+# clean up form
+# make role => is_admin
+# fix invite mail frappe logo
+
```
I guess these are already done (?) except `make role => is_admin`

s-aga-r commented on mail/mail/doctype/mail_account_request/mail_account_request.py:
```diff
@@ -107,12 +139,10 @@ def set_otp(self) -> None:
 	def send_verification_email(self) -> None:
 		"""Send verification email to the user."""
 
```
```suggestion
if not self.email:
    return # or throw
```

krantheman committed:
fix: multiple
- (Add Member): resest error after changing 'Send Invite'
- (Domain): show subdomain and root domain only if True
- (Mail Account Request): make email lowercase before insert

krantheman committed:
fix: validate tenant admin for get_members

s-aga-r committed:
chore: `linter`

s-aga-r committed:
refactor: make account email lowercase

s-aga-r committed:
chore: redirect `/login` to `/mail/login`

s-aga-r committed:
chore: redirect `/signup` to `/mail/signup`

s-aga-r committed:
fix(ux): use input type as `text`

krantheman committed:
refactor: redirect admin to desk

krantheman committed:
fix: routing

krantheman committed:
fix: routing

krantheman committed:
refactor: check if domain request already exists before creating new one

krantheman committed:
fix: check Mail Domain instead of Mail Domain Request for verified domains

krantheman committed:
feat: make domain name in error message bold

krantheman committed:
fix: hacky fix to redirect admin to app after setup wizard

s-aga-r committed:
fix: use remote url for frappe-logo

krantheman committed:
refactor(Mail Account Request): role => is_admin

krantheman committed:
feat(Members): show owner

s-aga-r commented on mail/mail/doctype/mail_account_request/mail_account_request.json:
```diff
@@ -161,11 +151,21 @@
+   "in_list_view": 1,
+   "in_standard_filter": 1,
+   "label": "Is Admin",
+   "reqd": 1,
```
remove reqd

s-aga-r commented on mail/mail/doctype/mail_tenant/mail_tenant.py:
```diff
@@ -26,13 +26,13 @@ def after_insert(self) -> None:
 		self.add_member(self.user)
 
-	def add_member(self, user: str, is_admin: bool = False) -> str:
+	def add_member(self, user: str, is_admin: Literal[0, 1] = 0) -> str:
```
Let's use bool everywhere. It looks much cleaner.

s-aga-r commented on mail/mail/doctype/mail_account/mail_account.py:
```diff
@@ -177,40 +177,50 @@ def _create_user_for_mail_account(
 	last_name: str | None = None,
 	password: str | None = None,
-	role: Literal["Mail User", "Mail Admin"] = "Mail User",
+	is_admin: Literal[0, 1] = 0,
```
bool

krantheman committed:
fix(Copy): remove scrollbar

krantheman committed:
fix(Members): don't show remove admin/remove member for owner

krantheman committed:
chore: change is_admin type annotation from literal to bool

krantheman committed:
refactor(Mail Account Request): clean up form flow

s-aga-r commented on mail/api/admin.py:
```diff
@@ -90,7 +90,7 @@ def add_member(
 	account_request.invited_by = frappe.session.user
 	account_request.email = email
-	account_request.send_email = send_invite
+	account_request.send_invite = send_invite
```
account_request.send_invite = cint(send_invite)

krantheman committed:
fix: type for checkbox field

s-aga-r committed:
fix: do not explicit validate email

s-aga-r committed:
fix: validate account email
